Added dnssrv lookup module.

rmkraus opened the pull request:
##### SUMMARY
This adds a new lookup module, dnssrv. It was modeled after the dnstxt lookup module, but for SRV records. It parses the results and returns them grouped and sorted per the RFC 2782 spec.

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
plugin/lookup/dnssrv

##### ADDITIONAL INFORMATION
Formerly, this same action could be mostly replicated using the dig lkp module. That code would look like this:
```yaml
record: '{{ ["target", "port"] | map("extract", lookup("dig", "_test._tcp.example.com/SRV", "flat=0", wantlist=True).0) | map("regex_replace", "\.$", "") | join(":") }}'
```

With the dnssrv lookup module, it now looks like this:
```yaml
record: '{{ lookup("dnssrv", "_test._tcp.example.com").0.0.target }}'
```

This lookup plugin also extends what is reasonably possible with the dig plugin as everything is sorted and provides the provided weighting information from the DNS records.

rmkraus committed:
Created dnssrv lookup plugin.

This lookup plugin operates in a very similar way as the dnstxt lookup
plugin, except it will lookup SRV records from DNS. It then parses the
records and returns them sorted and grouped per the RFC 2782 spec.

Signed-off-by: Ryan Kraus <rmkraus@gmail.com>

rmkraus committed:
Added tests for dnssrv lookup module.

1) Moved the code for the actual DNS query out of the run method in the
   LookupModule class. This was done so it could be mocked in tests.
2) Added two tests. One to test a good lookup and one to test a bad lookup.

Signed-off-by: Ryan Kraus <rmkraus@gmail.com>

rmkraus committed:
Fixed typo in dnssrv documentation.

Signed-off-by: Ryan Kraus <rmkraus@gmail.com>

rmkraus committed:
Resolved testing issues with dnssrv lookup plugin and test.

Signed-off-by: Ryan Kraus <rmkraus@gmail.com>

rmkraus committed:
Added dnspython dependency to unit tests.

This dependency is needed to test dnssrv.

Signed-off-by: Ryan Kraus <rmkraus@gmail.com>

rmkraus commented:
Sorry for the multiple commits. This is my first time playing with the Ansible unit tests. It looks like everything is passing now except two tests that failed because of unrelated web issues. Should be ready to review! Thanks.

rmkraus committed:
Added changelog fragment for PR #58806.

willthames commented:
Any reason we have to have a plugin for each DNS record type - why not just `lookup('dns', "_test._tcp.example.com", type='SRV')` (and make `dnstxt` and `dnssrv` aliases if need be?)

rmkraus commented:
The dig lookup module actually serves as a very good general purpose DNS lookup tool.

However, SRV records require special handling. To be used correctly, all of the records need to be parsed, sorted, and then converted into HOST:PORT combinations. To make things a bit more interesting, Python does not appreciate DNS names with the trailing dot (example.com.) and SRV records cannot exist without the trailing dot. If you look at unit tests for this lookup plugin, you can see some mock DNS records and what they will be converted to once they are parsed. 

Without this plugin, it is totally possible to pull an SRV record out and use it. That example is in the PR description. However, it is ugly, non-intuitive, and you have no guarantee you are even using the correct entry. You will simply use whatever one the DNS server decided to provide first. 

I don't think building this logic into the dig lookup plugin is best because the dig plugin is designed to return the raw DNS records, optionally converted to a dict. This plugin is meant to return data in a format that is well suited for use in automations. I think there is a use case for both. Also, building special logic into the dig plugin that is used only for one type of record seems counterintuitive.

This plugin is useful for enterprise environments to allow Ansible playbooks to simply search for LDAP/Kerberos/AD/NTP services for a given domain the way they are currently defined today, rather than requiring that data be duplicated in the inventory.

willthames commented on test/units/plugins/lookup/test_dnssrv.py:
```diff
@@ -0,0 +1,71 @@
+    rdata.from_text(IN, SRV, '20 40 8443 demo-4.example.com.')
+]
+MOCK_ANSWER = [
+    [{'priority': 10, 'weight': 3 / 6. * 100, 'target': 'demo-3.example.com:8443'},
```
not sure `6.` is needed as you have `from __future__ import division`

rmkraus replied:
I suppose not. Old habits die hard. I've removed the decimals.

rmkraus committed:
Removed unnecessary decimals in dnssrv lookup plugin unit tests.

Signed-off-by: Ryan Kraus <rmkraus@gmail.com>

rmkraus commented:
@willthames 

I think the tests just need to be rerun. The errors are all socket timeouts that have nothing to do with the code I changed. I removed that unnecessary decimal per your recommendation.

Akasurde commented:
Hi @rmkraus, Thank you very much for your interest in Ansible. This plugin/module is no longer maintained in this repository and has been migrated to https://github.com/ansible-collections/community.general<br><br>If you have further questions please stop by IRC or the mailing list:<br><br>* IRC: #ansible on irc.freenode.net<br>* mailing list: https://groups.google.com/forum/#!forum/ansible-project<br><br>needs_info<br>
